File: Go/Go_1000-/2384.go
// Go
// leetcode 2384
// https://leetcode.cn/problems/largest-palindromic-number/

func largestPalindromic(num string) string {
	// 统计每个数字出现次数
	cnt := make([]int , 10)
	for _, b := range num {
		cnt[int(b - '0')]++
	}

	// 成对的数录入
	tmp := make([]byte, 0, 4)
	for i := 9; i >= 0; i-- {
		if i == 0 && len(tmp) == 0 {
			continue
		}

		n := cnt[i] >> 1
		cnt[i] &= 1
		for n > 0 {
			tmp = append(tmp, byte(i + '0'))
			n--
		}
	}

	// 是否能在中间加一个数
	p := len(tmp) - 1
	for i := 9; i >= 0; i-- {
		if cnt[i] != 0 {
			tmp = append(tmp, byte(i + '0'))
			break
		}
	}

	// 填充尾巴
	for i := p; i >= 0; i-- {
		tmp = append(tmp, tmp[i])
	}

	return string(tmp)
}

File: cpp/cpp_1000-/2384.cc
// C++
// leetcode 2384
// https://leetcode.cn/problems/largest-palindromic-number/

class Solution {
public:
    string largestPalindromic(string num) {
        int cnt[10];
        memset(cnt, 0, sizeof(cnt));
        for (int i = 0; i < num.size(); i++)
        	cnt[num[i] - '0']++;

        vector<char> tmp;
        for (int i = 9; i >= 0; i--) {
        	if (i == 0 && tmp.size() == 0)
        		continue;
            tmp.insert(tmp.end(), cnt[i] >> 1, (char) ('0' + i));
        	cnt[i] &= 1;
        }

        int p = tmp.size() - 1;
        for (int i = 9; i >= 0; i--) {
        	if (cnt[i] != 0) {
        		tmp.push_back((char) ('0' + i));
        		break;
        	}
        }

        while (p >= 0) {
        	tmp.push_back(tmp[p--]);
        }

        return string(tmp.begin(), tmp.end());
    }
};